plugins/upower.c

```diff
@@ -0,0 +1,276 @@
+/*
+ *
+ *  oFono - Open Source Telephony
+ *
+ *  Copyright (C) 2015 Canonical Ltd.
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License version 2 as
+ *  published by the Free Software Foundation.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <glib.h>
+#include <gdbus.h>
+#include <ofono.h>
+
+#define OFONO_API_SUBJECT_TO_CHANGE
+#include <ofono/plugin.h>
+#include <ofono/modem.h>
+#include <ofono/log.h>
+#include <ofono/emulator.h>
+
+#define UPOWER_SERVICE            "org.freedesktop.UPower"
+#define UPOWER_DEVICE_INTERFACE   UPOWER_SERVICE ".Device"
+#define UPOWER_BASE_BATT_PATH     "/org/freedesktop/UPower/devices/battery_"
+#define UPOWER_TOUCH_BATT_PATH    UPOWER_BASE_BATT_PATH "battery"
```

Comment:
I didn't mean a match or something like that. But we could register to DeviceAdded/DeviceRemoved signals from the upower object and enumerate all objects on startup with looking at their properties.